File: server/server/stable_identity.py
"""Cross-runtime stable identities for persisted and API-visible entities."""

from __future__ import annotations

from hashlib import sha256
from typing import Mapping
from urllib.parse import urlsplit, urlunsplit


STABLE_IDENTITY_VERSION = "v1"

_VOLATILE_PLAYBACK_HEADERS = {
    "connection",
    "content-length",
    "if-range",
    "range",
}
_SENSITIVE_HEADER_NAMES = {
    "authorization",
    "cookie",
    "cookie2",
    "proxy-authorization",
    "x-api-key",
    "x-auth-token",
    "x-access-token",
    "access-token",
    "api-key",
}


def stable_digest(value: str) -> str:
    return sha256(value.encode("utf-8")).hexdigest()


def stable_subject_key(
    source: str,
    identifier: object,
    media_type: str | None = None,
) -> str:
    normalized_source = source.strip().lower()
    normalized_identifier = str(identifier).strip()
    if normalized_source == "bangumi":
        return f"bangumi:{normalized_identifier}"
    source_parts = normalized_source.split(":")
    if source_parts[0] == "bangumi" and len(source_parts) >= 2:
        return f"bangumi:{source_parts[1]}"
    if source_parts[0] == "tmdb":
        raw_type = (
            source_parts[1]
            if len(source_parts) >= 2
            else (media_type or "").strip().lower()
        )
        normalized_type = "tv" if raw_type == "series" else raw_type
        provider_identifier = (
            source_parts[2] if len(source_parts) >= 3 else normalized_identifier
        )
        if normalized_type in {"tv", "movie"} and provider_identifier:
            return f"tmdb:{normalized_type}:{provider_identifier}"
    if source_parts[0] == "archive" and len(source_parts) >= 2:
        return _generic_subject_key("archive", ":".join(source_parts[1:]))
    if source_parts[0] == "m3u-channel" and len(source_parts) >= 3:
        return _generic_subject_key(
            f"m3u-channel:{source_parts[1]}", ":".join(source_parts[2:])
        )
    if source_parts[0] == "cinemeta" and len(source_parts) >= 3:
        return _generic_subject_key(
            f"cinemeta:{source_parts[1]}", ":".join(source_parts[2:])
        )
    return _generic_subject_key(
        normalized_source,
        normalized_identifier,
        media_type,
    )


def _generic_subject_key(
    source: str,
    identifier: str,
    media_type: str | None = None,
) -> str:
    canonical = _frame_identity_parts(
        (
            source,
            identifier,
            (media_type or "").strip().lower(),
        )
    )
    digest = stable_digest(f"subject|{STABLE_IDENTITY_VERSION}|{canonical}")
    return f"subject:{STABLE_IDENTITY_VERSION}:{digest}"


def stable_episode_key(subject_key: str, normalized_number: object) -> str:
    return (
        f"{STABLE_IDENTITY_VERSION}|{subject_key.strip()}|"
        f"episode:{_normalize_number(normalized_number)}"
    )


def stable_playback_line_key(
    provider_id: str,
    episode_key: str,
    uri: str,
    headers: Mapping[str, str] | None = None,
) -> str:
    canonical = _frame_identity_parts(
        (
            provider_id.strip(),
            episode_key.strip(),
            canonical_identity_uri(uri),
            stable_header_fingerprint(headers or {}),
        )
    )
    digest = stable_digest(f"line|{STABLE_IDENTITY_VERSION}|{canonical}")
    return f"line:{STABLE_IDENTITY_VERSION}:{digest}"


def stable_download_task_key(
    subject_key: str,
    episode_key: str,
    provider_id: str = "",
) -> str:
    canonical = _frame_identity_parts(
        (subject_key.strip(), episode_key.strip(), provider_id.strip())
    )
    digest = stable_digest(f"download|{STABLE_IDENTITY_VERSION}|{canonical}")
    return f"download:{STABLE_IDENTITY_VERSION}:{digest}"


def stable_rule_key(
    rule_id: str,
    engine: str,
    source_repository: str = "",
    content_hash: str = "",
) -> str:
    canonical = _frame_identity_parts(
        (
            rule_id.strip(),
            engine.strip().lower(),
            _canonical_uri_or_text(source_repository),
            content_hash.strip().lower(),
        )
    )
    digest = stable_digest(f"rule|{STABLE_IDENTITY_VERSION}|{canonical}")
    return f"rule:{STABLE_IDENTITY_VERSION}:{digest}"


def stable_int63(value: str) -> int:
    result = int(stable_digest(value)[:16], 16) & ((1 << 63) - 1)
    return result or 1


def canonical_identity_uri(value: str) -> str:
    normalized = value.strip()
    parsed = urlsplit(normalized)
    if not parsed.scheme or not parsed.netloc or not parsed.hostname:
        raise ValueError("Identity URI must be absolute")
    scheme = parsed.scheme.lower()
    try:
        port = parsed.port
    except ValueError as error:
        raise ValueError("Identity URI contains an invalid port") from error
    if (scheme == "http" and port == 80) or (scheme == "https" and port == 443):
        port = None
    user_info = ""
    if "@" in parsed.netloc:
        user_info = parsed.netloc.rsplit("@", 1)[0] + "@"
    host = parsed.hostname.lower()
    if ":" in host and not host.startswith("["):
        host = f"[{host}]"
    netloc = f"{user_info}{host}{f':{port}' if port is not None else ''}"
    return urlunsplit((scheme, netloc, parsed.path, parsed.query, ""))


def stable_header_fingerprint(headers: Mapping[str, str]) -> str:
    entries = sorted(
        ((str(name), str(value)) for name, value in headers.items()),
        key=lambda item: (item[0].lower(), item[0], item[1]),
    )
    normalized: dict[str, str] = {}
    for raw_name, raw_value in entries:
        name = raw_name.strip().lower()
        value = raw_value.strip()
        if not name or not value or name in _VOLATILE_PLAYBACK_HEADERS:
            continue
        if name in {"referer", "origin"}:
            value = _canonical_uri_or_text(value)
        elif _is_sensitive_header(name):
            value = f"sha256:{stable_digest(value)}"
        normalized[name] = value
    canonical = _frame_identity_parts(
        f"{name}={normalized[name]}" for name in sorted(normalized)
    )
    return stable_digest(f"headers|{STABLE_IDENTITY_VERSION}|{canonical}")


def _is_sensitive_header(name: str) -> bool:
    return (
        name in _SENSITIVE_HEADER_NAMES
        or "token" in name
        or "secret" in name
        or name.endswith("-key")
    )


def _canonical_uri_or_text(value: str) -> str:
    normalized = value.strip()
    if not normalized:
        return ""
    try:
        return canonical_identity_uri(normalized)
    except ValueError:
        return normalized


def _frame_identity_parts(parts) -> str:
    return "|".join(f"{len(part.encode('utf-8'))}:{part}" for part in parts)


def _normalize_number(value: object) -> str:
    raw = str(value).strip()
    import re

    match = re.fullmatch(r"([+-]?)(\d+)(?:\.(\d+))?", raw)
    if match is None:
        return raw.lower()
    integer = match.group(2).lstrip("0") or "0"
    fraction = (match.group(3) or "").rstrip("0")
    is_zero = integer == "0" and not fraction
    sign = "-" if match.group(1) == "-" and not is_zero else ""
    return f"{sign}{integer}{f'.{fraction}' if fraction else ''}"

File: lib/src/core/identity/stable_identity.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

const stableIdentityVersion = 'v1';

const _volatilePlaybackHeaders = <String>{
  'connection',
  'content-length',
  'if-range',
  'range',
};

const _sensitiveHeaderNames = <String>{
  'authorization',
  'cookie',
  'cookie2',
  'proxy-authorization',
  'x-api-key',
  'x-auth-token',
  'x-access-token',
  'access-token',
  'api-key',
};

String stableDigest(String value) =>
    sha256.convert(utf8.encode(value)).toString();

String stableSubjectKey({
  required String source,
  required Object identifier,
  String? mediaType,
}) {
  final normalizedSource = source.trim().toLowerCase();
  final normalizedIdentifier = identifier.toString().trim();
  if (normalizedSource == 'bangumi') {
    return 'bangumi:$normalizedIdentifier';
  }
  final sourceParts = normalizedSource.split(':');
  if (sourceParts.firstOrNull == 'bangumi' && sourceParts.length >= 2) {
    return 'bangumi:${sourceParts[1]}';
  }
  if (sourceParts.firstOrNull == 'tmdb') {
    final rawType = sourceParts.length >= 2
        ? sourceParts[1]
        : (mediaType ?? '').trim().toLowerCase();
    final normalizedType = rawType == 'series' ? 'tv' : rawType;
    final providerIdentifier = sourceParts.length >= 3
        ? sourceParts[2]
        : normalizedIdentifier;
    if ((normalizedType == 'tv' || normalizedType == 'movie') &&
        providerIdentifier.isNotEmpty) {
      return 'tmdb:$normalizedType:$providerIdentifier';
    }
  }
  if (sourceParts.firstOrNull == 'archive' && sourceParts.length >= 2) {
    return _genericSubjectKey(
      source: 'archive',
      identifier: sourceParts.skip(1).join(':'),
    );
  }
  if (sourceParts.firstOrNull == 'm3u-channel' && sourceParts.length >= 3) {
    return _genericSubjectKey(
      source: 'm3u-channel:${sourceParts[1]}',
      identifier: sourceParts.skip(2).join(':'),
    );
  }
  if (sourceParts.firstOrNull == 'cinemeta' && sourceParts.length >= 3) {
    return _genericSubjectKey(
      source: 'cinemeta:${sourceParts[1]}',
      identifier: sourceParts.skip(2).join(':'),
    );
  }
  return _genericSubjectKey(
    source: normalizedSource,
    identifier: normalizedIdentifier,
    mediaType: mediaType,
  );
}

String _genericSubjectKey({
  required String source,
  required String identifier,
  String? mediaType,
}) {
  final canonical = _frameIdentityParts([
    source,
    identifier,
    (mediaType ?? '').trim().toLowerCase(),
  ]);
  return 'subject:$stableIdentityVersion:${stableDigest('subject|$stableIdentityVersion|$canonical')}';
}

String stableEpisodeKey({
  required String subjectKey,
  required Object normalizedNumber,
}) =>
    '$stableIdentityVersion|${subjectKey.trim()}|episode:${_normalizeNumber(normalizedNumber)}';

String stablePlaybackLineKey({
  required String providerId,
  required String episodeKey,
  required String uri,
  Map<String, String> headers = const {},
}) {
  final canonical = _frameIdentityParts([
    providerId.trim(),
    episodeKey.trim(),
    canonicalIdentityUri(uri),
    stableHeaderFingerprint(headers),
  ]);
  return 'line:$stableIdentityVersion:${stableDigest('line|$stableIdentityVersion|$canonical')}';
}

String stableDownloadTaskKey({
  required String subjectKey,
  required String episodeKey,
  String providerId = '',
}) {
  final canonical = _frameIdentityParts([
    subjectKey.trim(),
    episodeKey.trim(),
    providerId.trim(),
  ]);
  return 'download:$stableIdentityVersion:${stableDigest('download|$stableIdentityVersion|$canonical')}';
}

String stableRuleKey({
  required String ruleId,
  required String engine,
  String sourceRepository = '',
  String contentHash = '',
}) {
  final repository = _canonicalUriOrText(sourceRepository);
  final canonical = _frameIdentityParts([
    ruleId.trim(),
    engine.trim().toLowerCase(),
    repository,
    contentHash.trim().toLowerCase(),
  ]);
  return 'rule:$stableIdentityVersion:${stableDigest('rule|$stableIdentityVersion|$canonical')}';
}

int stableInt63(String value) {
  final prefix = stableDigest(value).substring(0, 16);
  final parsed = BigInt.parse(prefix, radix: 16);
  final result = (parsed & ((BigInt.one << 63) - BigInt.one)).toInt();
  return result == 0 ? 1 : result;
}

String canonicalIdentityUri(String value) {
  final uri = Uri.tryParse(value.trim());
  if (uri == null || !uri.hasScheme || !uri.hasAuthority || uri.host.isEmpty) {
    throw FormatException('Identity URI must be absolute.', value);
  }
  final scheme = uri.scheme.toLowerCase();
  final defaultPort =
      (scheme == 'http' && uri.port == 80) ||
      (scheme == 'https' && uri.port == 443);
  final userInfo = uri.userInfo.isEmpty ? '' : '${uri.userInfo}@';
  final normalizedHost = uri.host.toLowerCase();
  final host = normalizedHost.contains(':')
      ? '[$normalizedHost]'
      : normalizedHost;
  final port = !defaultPort && uri.hasPort ? ':${uri.port}' : '';
  final query = uri.hasQuery ? '?${uri.query}' : '';
  return '$scheme://$userInfo$host$port${uri.path}$query';
}

String stableHeaderFingerprint(Map<String, String> headers) {
  final entries = headers.entries.toList(growable: false)
    ..sort((left, right) {
      final byName = left.key.toLowerCase().compareTo(right.key.toLowerCase());
      if (byName != 0) return byName;
      final byOriginalName = left.key.compareTo(right.key);
      if (byOriginalName != 0) return byOriginalName;
      return left.value.compareTo(right.value);
    });
  final normalized = <String, String>{};
  for (final entry in entries) {
    final name = entry.key.trim().toLowerCase();
    final rawValue = entry.value.trim();
    if (name.isEmpty ||
        rawValue.isEmpty ||
        _volatilePlaybackHeaders.contains(name)) {
      continue;
    }
    final value = switch (name) {
      'referer' || 'origin' => _canonicalUriOrText(rawValue),
      _ when _isSensitiveHeader(name) => 'sha256:${stableDigest(rawValue)}',
      _ => rawValue,
    };
    normalized[name] = value;
  }
  final canonical = _frameIdentityParts([
    for (final name in normalized.keys.toList()..sort())
      '$name=${normalized[name]}',
  ]);
  return stableDigest('headers|$stableIdentityVersion|$canonical');
}

bool _isSensitiveHeader(String name) =>
    _sensitiveHeaderNames.contains(name) ||
    name.contains('token') ||
    name.contains('secret') ||
    name.endsWith('-key');

String _canonicalUriOrText(String value) {
  final normalized = value.trim();
  if (normalized.isEmpty) return '';
  try {
    return canonicalIdentityUri(normalized);
  } on FormatException {
    return normalized;
  }
}

String _frameIdentityParts(Iterable<String> parts) =>
    parts.map((part) => '${utf8.encode(part).length}:$part').join('|');

String _normalizeNumber(Object value) {
  final raw = value.toString().trim();
  final match = RegExp(r'^([+-]?)(\d+)(?:\.(\d+))?$').firstMatch(raw);
  if (match == null) return raw.toLowerCase();
  final integer = match.group(2)!.replaceFirst(RegExp(r'^0+(?=\d)'), '');
  final fraction = (match.group(3) ?? '').replaceFirst(RegExp(r'0+$'), '');
  final isZero = int.tryParse(integer) == 0 && fraction.isEmpty;
  final sign = match.group(1) == '-' && !isZero ? '-' : '';
  return fraction.isEmpty ? '$sign$integer' : '$sign$integer.$fraction';
}
